### PR TITLE
[Scala client bindings] Fix a "deadlock" in CommandRetryFlow

### PR DESCRIPTION
**Pull request by gerolf-da**
The "deadlock" occurs when a command is sent to the RETRY_PORT, but at
the same time new commands are coming in as well. If the
commandSubmissionFlow is limited (e.g. via the MaxInFlight flow), then
backpressure is exerted on the merge flow and ultimately back to the
retryDecider. This in turn prevents new command results from being
propagated to the PROPAGATE_PORT, and therefore the limited
commandSubmissionFlow ends up backpressuring itself into a "deadlock".

To circumvent this, two new measures are in place:
1. The Merge flow is now a MergePreferred flow, giving precedence to
submission retries.
2. An additional buffer is added to the retry feedback loop, so that the
limited commandSubmissionFlow can keep making progress.

Ideally this buffer would have an additional feature of rerouting new
elements to a completely different port if the buffer is full, but that
would likely require a custom graph stage.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: language-support/scala/bindings-akka/src/main/scala/com/digitalasset/ledger/client/binding/retrying/CommandRetryFlow.scala
Comment (cocreature) [inline language-support/scala/bindings-akka/src/main/scala/com/digitalasset/ledger/client/binding/retrying/CommandRetryFlow.scala:113]: So if I understand you correctly we still deadlock if this buffer is full?

Comment (gerolf-da) [reply language-support/scala/bindings-akka/src/main/scala/com/digitalasset/ledger/client/binding/retrying/CommandRetryFlow.scala:113]: That's right, but together with the MergePreferred flow this should be dramatically reduced.
Apologies, I should have been explicit about it.

**Review by cocreature — APPROVED** [review]
Looks sensible, very nice work tracking this down!

I admit I’m slightly scared by the complexity of the retry flow especially since it only seems to be getting more and more complex but I don’t have a great suggestion for simplifying it either unfortunately.

I recommend waiting for a second review from someone who is a bit more of an akka streams expert than I am.

File: language-support/scala/bindings-akka/src/main/scala/com/digitalasset/ledger/client/binding/retrying/CommandRetryFlow.scala
Comment (S11001001) [inline language-support/scala/bindings-akka/src/main/scala/com/digitalasset/ledger/client/binding/retrying/CommandRetryFlow.scala:118]: This issue existed before, but is exacerbated by the buffer:

1. What happens if there are commands in the buffer when a `CommandRetryFlow`'s upstream completes? (I believe this code implies "they are dropped", but could be mistaken.)
2. Is that behavior desirable? Is it acceptable?

**Review by S11001001 — APPROVED** [review]

File: language-support/scala/bindings-akka/src/main/scala/com/digitalasset/ledger/client/binding/retrying/CommandRetryFlow.scala
Comment (gerolf-da) [reply language-support/scala/bindings-akka/src/main/scala/com/digitalasset/ledger/client/binding/retrying/CommandRetryFlow.scala:118]: If the upstream completes, then the completion signal is propagated downstream and the retries are lost. Indeed, the buffer exacerbates that loss.

**Comment by gerolf-da** [comment]
Thank you for your comments, @cocreature and @S11001001.
I think both are valid concerns.
Since the deadlock is technically still possible, do you want to track this separately?
Regarding the lost retries on upstream completion: I'd need to think a bit more about how to solve this, but I think this is a tricky one.

**Comment by cocreature** [comment]
> Since the deadlock is technically still possible, do you want to track this separately?

opened #10113 and #10114

**Comment by gerolf-da** [comment]
https://github.com/digital-asset/daml/pull/10111/commits/307b399dd8d48e0f689bff2796a11515b817f32c moves the `MaxInFlight` stage outside the `CommandRetryFlow`, which means there will be no backpressure internally to `CommandRetryFlow` and the deadlock will be prevented.

**Review by stefanobaghino-da — APPROVED** [review]
Nothing to add on top of what has already been mentioned.

I don't think this is a viable approach, but purely in the interest of putting the idea out there for consideration: what would be the impact of dropping retries entirely and ask the clients to use an explicit retry policy of their own?

**Comment by gerolf-da** [comment]
@stefanobaghino-da: this is actually a very sensible suggestion. The place that currently uses `LedgerClientBinding#retryingConfirmedCommands` could also use `LedgerClientBinding#commands` and let the application handle the errors. I will check with if this works for the application in question.

**Comment by gerolf-da** [comment]
Closing this stale PR.

**Pull request closed without merging**
